Backport 2.x: ssl-opt.sh: Skip tests instead of conditional hard abort

yutotakano opened the pull request:
This is a relatively straightforward backport of https://github.com/ARMmbed/mbedtls/pull/4695 for 2.x.

Fixes #4471.

yutotakano commented:
Rebased upon the top of development_2.x and force-pushed. It now contains all changes post-review from #4695 (remove incorrect comments, unneeded echos, add requires_config_value_equals, etc).

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -4608,6 +4641,7 @@ run_test    "Authentication, CA callback: client max_int+1 chain, server require
+requires_config_value_equals "MBEDTLS_X509_MAX_INTERMEDIATE_CA" $MAX_IM_CA
 requires_full_size_output_buffer
 requires_config_enabled MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK
 run_test    "Authentication, CA callback: client max_int chain, server required" \
```
Missing `requires_max_content_len 4096` on the test case `"SSLv3 with extensions, server side"`.

I found this by checking that each use of `max_frag_len` has a corresponding `requires_max_content_len` (or doesn't need one). This check needs to be done for each Mbed TLS versions since different versions have different test cases. Here there's an extra test case in 2.2x and 2.16 for a feature that was removed in 3.0.

gilles-peskine-arm reviewed (changes requested):
Some test cases are missing `requires_config_value_equals` (see also https://github.com/ARMmbed/mbedtls/pull/4695#pullrequestreview-699376527).

yutotakano commented:
> Some test cases are missing `requires_config_value_equals`

Sorry, did you mean `requires_max_content_len` as written on the comment below, or was there somewhere else I missed? All the Authentication tests (not just max_int) pass/skip properly with values other than 8.

gilles-peskine-arm commented:
Yes, sorry, I meant `requires_max_content_len` as detailed in individual comments.

yutotakano commented:
Cherry-picked commits from `fix-ssl-opt.sh-hard-abort`, addressed the SSLv3 test case missing constraints, and rebased upon current state of `development_2.x`. Ready for review!

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -6927,6 +6966,7 @@ run_test    "Large client packet TLS 1.2 AEAD shorter tag" \
             -s "Read from client: $MAX_CONTENT_LEN bytes read"
 
 # Test for large server packets
+# The tests below fail when the server's OUT_CONTENT_LEN is less than 16384.
```
So why is there no corresponding `requires`?

yutotakano replied:
I believe they would be out of scope for this PR. My reasoning is based on [this discussion](https://github.com/ARMmbed/mbedtls/pull/4695#discussion_r662995615), where we concluded that if it's a pre-existing issue (which this is), then it should be in a different PR. 

Before this PR, setting OUT_CONTENT_LEN to e.g. 8096 would not trigger the hard abort but still fail these tests.

The note was added for future reference. Feel free to let me know if this should be included in the scope as the line is definitely a little blurry.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -8617,6 +8683,7 @@ requires_config_enabled MBEDTLS_RSA_C
 requires_config_enabled MBEDTLS_ECDSA_C
 requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
 requires_gnutls
+requires_max_content_len 2048
```
The DTLS 1.0 test case just below should have the same requirement.

Also goes for the client and openssl test cases below, where similarly the 1.2 version have a requirement but the 1.0 version doesn't.

gilles-peskine-arm reviewed (changes requested):
There's one inconsistency left, other than that looks good to me.
